File: design/static/validation.js
/**
 * 与 src/system/validation.py 保持一致的客户端校验。
 * 常量由页面模板注入到 window.DETECT_VALIDATION。
 */
(function (global) {
  const DEFAULTS = {
    minTextLen: 6,
    maxTextLen: 10000,
    minTokenCount: 1,
    maxSpecialRatio: 0.85,
  };

  const PUNCTUATIONS = new Set(
    "，。！？；：、“”‘’（）()《》<>【】[],.!?;:'\"-".split("")
  );
  const TOKEN_RE = /[\u4e00-\u9fffA-Za-z0-9]+/g;
  const CONTROL_RE = /[\x00-\x08\x0b\x0c\x0e-\x1f\x7f-\x9f]/;

  function cfg() {
    return { ...DEFAULTS, ...(global.DETECT_VALIDATION || {}) };
  }

  function cleanText(text) {
    return text.trim().replace(/\u3000/g, " ").replace(/\s+/g, " ");
  }

  function countSpecial(text) {
    let n = 0;
    for (const c of text) {
      if (!c.match(/[\u4e00-\u9fffA-Za-z0-9]/) && !PUNCTUATIONS.has(c) && !/\s/.test(c)) {
        n += 1;
      }
    }
    return n;
  }

  function validateDetectText(raw) {
    const { minTextLen, maxTextLen, minTokenCount, maxSpecialRatio } = cfg();

    if (raw == null) {
      return { ok: false, text: "", error: "请输入待检测文本。" };
    }
    if (typeof raw !== "string") {
      return { ok: false, text: "", error: "text 字段必须是字符串。" };
    }
    if (raw.includes("\x00")) {
      return { ok: false, text: "", error: "文本包含非法空字符。" };
    }
    if (CONTROL_RE.test(raw)) {
      return { ok: false, text: "", error: "文本包含非法控制字符。" };
    }

    const text = cleanText(raw);
    if (!text) {
      return { ok: false, text: "", error: "请输入待检测文本。" };
    }

    const length = text.length;
    if (length < minTextLen) {
      return {
        ok: false,
        text: "",
        error: `文本过短，至少需要 ${minTextLen} 个字符（当前 ${length} 个）。`,
      };
    }
    if (length > maxTextLen) {
      return {
        ok: false,
        text: "",
        error: `文本过长，最多 ${maxTextLen} 个字符（当前 ${length} 个）。`,
      };
    }

    const tokens = text.match(TOKEN_RE) || [];
    if (tokens.length < minTokenCount) {
      return {
        ok: false,
        text: "",
        error: `有效内容不足，至少需包含 ${minTokenCount} 个中文/英文/数字词。`,
      };
    }

    if (countSpecial(text) / length > maxSpecialRatio) {
      return { ok: false, text: "", error: "特殊字符占比过高，请输入正常文本。" };
    }

    return { ok: true, text, error: "" };
  }

  global.validateDetectText = validateDetectText;
})(window);

File: design/src/system/validation.py
from __future__ import annotations

"""检测接口输入校验：与训练预处理规则对齐，并限制异常输入。"""

import re
from dataclasses import dataclass

from src.data.preprocess import clean_text
from src.features.extractor import PUNCTUATIONS, _tokens

# 与 preprocess.load_and_merge 中 `str.len() > 5` 保持一致。
MIN_TEXT_LEN = 6
MAX_TEXT_LEN = 10_000
MIN_TOKEN_COUNT = 1
MAX_SPECIAL_RATIO = 0.85

# 允许常见空白；其余 C0/C1 控制符拒绝。
_CONTROL_RE = re.compile(r"[\x00-\x08\x0b\x0c\x0e-\x1f\x7f-\x9f]")


@dataclass(frozen=True)
class TextValidationResult:
    ok: bool
    text: str = ""
    error: str = ""


def _count_special(text: str) -> int:
    return sum(1 for c in text if (not c.isalnum() and c not in PUNCTUATIONS and not c.isspace()))


def validate_detect_text(raw: object) -> TextValidationResult:
    """校验待检测文本，成功时返回清洗后的文本。"""

    if raw is None:
        return TextValidationResult(False, error="请输入待检测文本。")
    if not isinstance(raw, str):
        return TextValidationResult(False, error="text 字段必须是字符串。")

    if "\x00" in raw:
        return TextValidationResult(False, error="文本包含非法空字符。")
    if _CONTROL_RE.search(raw):
        return TextValidationResult(False, error="文本包含非法控制字符。")

    text = clean_text(raw)
    if not text:
        return TextValidationResult(False, error="请输入待检测文本。")

    length = len(text)
    if length < MIN_TEXT_LEN:
        return TextValidationResult(
            False,
            error=f"文本过短，至少需要 {MIN_TEXT_LEN} 个字符（当前 {length} 个）。",
        )
    if length > MAX_TEXT_LEN:
        return TextValidationResult(
            False,
            error=f"文本过长，最多 {MAX_TEXT_LEN} 个字符（当前 {length} 个）。",
        )

    tokens = _tokens(text)
    if len(tokens) < MIN_TOKEN_COUNT:
        return TextValidationResult(
            False,
            error=f"有效内容不足，至少需包含 {MIN_TOKEN_COUNT} 个中文/英文/数字词。",
        )

    special_ratio = _count_special(text) / length
    if special_ratio > MAX_SPECIAL_RATIO:
        return TextValidationResult(False, error="特殊字符占比过高，请输入正常文本。")

    return TextValidationResult(True, text=text)
